File: backend/src/main/java/com/backend/domain/evaluation/service/impl/OpenAiGateway.java
package com.backend.domain.evaluation.service.impl;

import com.backend.domain.evaluation.service.AiGateway;
import com.openai.client.OpenAIClient;
import com.openai.models.ChatModel;
import com.openai.models.responses.*;
import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression; // ★ import
import org.springframework.stereotype.Service;

import java.lang.reflect.Method;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
@ConditionalOnExpression("T(org.springframework.util.StringUtils).hasText(environment['openai.api.key'])")
@RequiredArgsConstructor
public class OpenAiGateway implements AiGateway {

    private final OpenAIClient client;

    @PostConstruct
    void logActive() { System.out.println("[AiGateway] OpenAiGateway ACTIVE"); }

    @Override
    public String complete(String content, String prompt) {
        String input = """
                [SYSTEM PROMPT]
                %s

                [USER CONTENT]
                %s

                [RESPONSE RULE]
                - 항상 한국어 텍스트만 출력하세요.
                """.formatted(
                prompt == null ? "" : prompt,
                content == null ? "" : content
        );

        var params = ResponseCreateParams.builder()
                .model(ChatModel.GPT_5_NANO)
                .input(input)
                .build();

        Response res = client.responses().create(params);

        try {
            Method m = res.getClass().getMethod("outputText");
            Object out = m.invoke(res);
            if (out instanceof String s && !s.trim().isEmpty()) return s.trim();
        } catch (Throwable ignore) {}

        String joined = extractOutputText(res);
        if (!joined.isEmpty()) return joined;

        return String.valueOf(res);
    }

    private static String extractOutputText(Response res) {
        if (res == null || res.output() == null) return "";
        return res.output().stream()
                .map(ResponseOutputItem::message)
                .flatMap(Optional::stream)
                .flatMap(msg -> {
                    List<ResponseOutputMessage.Content> cs = msg.content();
                    return (cs == null) ? Stream.<ResponseOutputMessage.Content>empty() : cs.stream();
                })
                .map(c -> {
                    try {
                        ResponseOutputText t = c.asOutputText();
                        return (t != null) ? t.text() : null;
                    } catch (Exception ignore) { return null; }
                })
                .filter(Objects::nonNull)
                .map(String::trim)
                .filter(s -> !s.isEmpty())
                .collect(Collectors.joining());
    }
}

File: backend/src/main/java/com/backend/domain/evaluation/service/impl/OpenAiGateway.kt
package com.backend.domain.evaluation.service.impl

import com.backend.domain.evaluation.service.AiGateway
import com.openai.client.OpenAIClient
import com.openai.models.ChatModel
import com.openai.models.responses.*
import jakarta.annotation.PostConstruct
import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression
import org.springframework.stereotype.Service
import java.lang.reflect.Method
import java.util.Objects
import java.util.Optional
import java.util.stream.Collectors
import java.util.stream.Stream

@Service
@ConditionalOnExpression("T(org.springframework.util.StringUtils).hasText(environment['openai.api.key'])")
class OpenAiGateway(
    private val client: OpenAIClient,
) : AiGateway {

    @PostConstruct
    fun logActive() {
        println("[AiGateway] OpenAiGateway ACTIVE")
    }

    override fun complete(content: String, prompt: String): String {
        val input = """
            [SYSTEM PROMPT]
            $prompt

            [USER CONTENT]
            $content

            [RESPONSE RULE]
            - 항상 한국어 텍스트만 출력하세요.
        """.trimIndent()

        val params = ResponseCreateParams.builder()
            .model(ChatModel.GPT_5_NANO)
            .input(input)
            .build()

        val res: Response = client.responses().create(params)

        // 1. 새로운 outputText() 헬퍼가 있을 경우 반영 (SDK 버전 대응)
        try {
            val m: Method = res.javaClass.getMethod("outputText")
            val out: Any? = m.invoke(res)
            if (out is String && out.trim().isNotEmpty()) {
                return out.trim()
            }
        } catch (_: Throwable) {
            // 무시하고 아래 fallback 로직 사용
        }

        // 2. responses().create() 기본 output 파싱
        val joined = extractOutputText(res)
        if (joined.isNotEmpty()) return joined

        // 3. 마지막 fallback
        return res.toString()
    }

    /**
     * OpenAI Response 객체에서 텍스트만 뽑아서 하나의 문자열로 합친다.
     */
    private fun extractOutputText(res: Response?): String {
        if (res == null) return ""

        val output = res.output() ?: return ""

        return output.stream()
            .map(ResponseOutputItem::message)
            .flatMap { opt: Optional<ResponseOutputMessage> -> opt.stream() }
            .flatMap { msg ->
                val cs = msg.content()
                if (cs == null) {
                    Stream.empty<ResponseOutputMessage.Content>()
                } else {
                    cs.stream()
                }
            }
            .map { c ->
                try {
                    val t: ResponseOutputText? = c.asOutputText()
                    t?.text()
                } catch (_: Exception) {
                    null
                }
            }
            .filter(Objects::nonNull)                     // null 제거
            .map { (it as String).trim() }                // ✅ 명시적으로 String 캐스팅 후 trim
            .filter { it.isNotEmpty() }
            .collect(Collectors.joining())
    }
}
